File: engine/julia/all_weather_vomma_engine.jl
# Module BE5 (Julia): Karl Domm All-Weather Options Portfolio & Tail Risk Vomma Engine
# Quantitative evaluation of SPAN margin slicing, 4 market regimes, and 5-delta teenie put positive vomma in Julia.

module AllWeatherVommaEngine

export classify_regime, audit_margin, evaluate_teenies

function classify_regime(spx_return::Float64, vix_spike::Float64)
    if vix_spike >= 35.0
        return "CRASH_MARKET"
    elseif spx_return < -8.0 && vix_spike < 30.0
        return "GRIND_DOWN_MARKET"
    elseif abs(spx_return) <= 4.0
        return "SIDEWAYS_MARKET"
    else
        return "RISING_BULL_MARKET"
    end
end

function audit_margin(pnl_12_down::Float64, pnl_20_down::Float64, pnl_10_up::Float64, capital::Float64)
    s12 = abs(min(0.0, pnl_12_down))
    s20 = abs(min(0.0, pnl_20_down)) / 2.0
    s10 = abs(min(0.0, pnl_10_up))

    req = max(s12, max(s20, s10))
    util = (req / max(1.0, capital)) * 100.0
    safe = util <= 65.0

    return (req = req, util = util, is_safe = safe)
end

function evaluate_teenies(core_vomma::Float64, num_teenies::Int)
    net_vomma = core_vomma + (num_teenies * 0.08)
    pos_vomma = net_vomma > 0.0
    return (net_vomma = net_vomma, has_positive_vomma = pos_vomma)
end

end

File: engine/rust/src/all_weather_vomma_engine.rs
//! Module BE4 (Rust): Karl Domm All-Weather Options Portfolio & Tail Risk Vomma Engine.
//! SIMD-optimized SPAN margin slicing, 4 market regimes, and 5-delta teenie put positive vomma calculations.

#[repr(C, align(64))]
#[derive(Debug, Clone, Copy)]
pub struct AllWeatherVommaState {
    pub worst_case_margin_req: f64,
    pub planned_capital: f64,
    pub margin_utilization_pct: f64,
    pub net_portfolio_vomma: f64,
    pub market_regime_id: u32,
    pub is_margin_safe: u32,
    pub has_positive_vomma: u32,
    pub _padding: [u8; 20],
}

impl AllWeatherVommaState {
    pub fn audit(
        pnl_12_down: f64,
        pnl_20_down: f64,
        pnl_10_up: f64,
        capital: f64,
        vix_spike: f64,
        core_vomma: f64,
        num_teenies: i32
    ) -> Self {
        let s12 = pnl_12_down.min(0.0).abs();
        let s20 = (pnl_20_down.min(0.0).abs()) / 2.0;
        let s10 = pnl_10_up.min(0.0).abs();

        let req = s12.max(s20.max(s10));
        let util = (req / capital.max(1.0)) * 100.0;
        let safe = if util <= 65.0 { 1 } else { 0 };

        let regime = if vix_spike >= 35.0 { 4 } else { 1 };
        let net_vomma = core_vomma + (num_teenies as f64 * 0.08);
        let pos_vomma = if net_vomma > 0.0 { 1 } else { 0 };

        Self {
            worst_case_margin_req: req,
            planned_capital: capital,
            margin_utilization_pct: util,
            net_portfolio_vomma: net_vomma,
            market_regime_id: regime,
            is_margin_safe: safe,
            has_positive_vomma: pos_vomma,
            _padding: [0u8; 20],
        }
    }
}
